fili-core/src/test/groovy/com/yahoo/bard/webservice/web/responseprocessors/EtagCacheResponseProcessorSpec.groovy

```diff
@@ -0,0 +1,106 @@
+import static javax.ws.rs.core.Response.Status.NOT_MODIFIED
+import static javax.ws.rs.core.Response.Status.OK
+
+import com.yahoo.bard.webservice.data.cache.TupleDataCache
+import com.yahoo.bard.webservice.druid.client.HttpErrorCallback
+import com.yahoo.bard.webservice.druid.model.query.DruidAggregationQuery
+import com.yahoo.bard.webservice.web.ErrorMessageFormat
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module
+
+import spock.lang.Specification
+
+class EtagCacheResponseProcessorSpec extends Specification {
+    private static final ObjectMapper MAPPER = new ObjectMapper()
+            .registerModule(new Jdk8Module().configureAbsentsAsNulls(false))
+    private static final int INTERNAL_SERVER_ERROR_STATUS_CODE = INTERNAL_SERVER_ERROR.getStatusCode()
+    private static final String CACHE_KEY = "cacheKey"
+
+    HttpErrorCallback httpErrorCallback
+    DruidAggregationQuery druidAggregationQuery
+    ResponseProcessor next
+    TupleDataCache<String, String, String> dataCache
+    EtagCacheResponseProcessor etagCacheResponseProcessor
+
+    def setup() {
+        druidAggregationQuery = Mock(DruidAggregationQuery)
+        httpErrorCallback = Mock(HttpErrorCallback)
+        next = Mock(ResponseProcessor)
+        next.getErrorCallback(druidAggregationQuery) >> httpErrorCallback
+        dataCache = Mock(TupleDataCache)
+        etagCacheResponseProcessor = new EtagCacheResponseProcessor(
+                next,
+                CACHE_KEY,
+                dataCache,
+                MAPPER
+        )
+    }
+
+    def "processResponse reports error when status code is missing"() {
+        given:
+        JsonNode json = Mock(JsonNode)
+        json.has(DruidJsonResponseContentKeys.STATUS_CODE.getName()) >> false
+
+        when:
+        etagCacheResponseProcessor.processResponse(json, druidAggregationQuery, Mock(LoggingContext))
+
+        then:
+        1 * httpErrorCallback.dispatch(
+                INTERNAL_SERVER_ERROR_STATUS_CODE,
+                ErrorMessageFormat.INTERNAL_SERVER_ERROR_REASON_PHRASE.format(),
+                ErrorMessageFormat.STATUS_CODE_MISSING_FROM_RESPONSE.format()
+        )
+    }
+
+    def "processResponse reads response from cache when status code is NOT_MODIFIED"() {
+        given:
+        JsonNode json = MAPPER.readTree(
+                """{"${DruidJsonResponseContentKeys.STATUS_CODE.name}": $NOT_MODIFIED.statusCode}"""
+        )
+        dataCache.getDataValue(CACHE_KEY) >> '[{"k1":"v1"}]'
+
+        when:
+        etagCacheResponseProcessor.processResponse(json, druidAggregationQuery, Mock(LoggingContext))
+
+        then:
+        json.get(DruidJsonResponseContentKeys.RESPONSE.getName()) == MAPPER.readTree('[{"k1":"v1"}]')
+    }
+
+    def "processResponse reports error when status code is OK but etag is missing and processResponse moves on without caching"() {
```

Comment:
I forgot to update test name. It keeps going without cache without reporting error